File: Kalk/model/espressione.cpp
#include "espressione.h"
#include "aritmeticresult.h"

Espressione::Espressione(const std::string s) : text(s), result(nullptr) {}

Espressione::~Espressione() {
    delete result;
}

Result* Espressione::getResult() const
{
    if(result)
        return result;
    return nullptr;
}

std::string Espressione::getText() const
{
    return std::string(text);
}

bool Espressione::checkParentesi() const
{
    int check = 0;
    for(unsigned int i = 0; i < text.length(); ++i)
        if(text[i] == '(')
            ++check;
        else if(text[i] == ')' && --check < 0)
            return 0;
    if(check != 0)
        return false;
    for(unsigned int i = 0; i < text.length() - 1; ++i)
        if (text[i] == '(' && text[i+1] == ')'  )
            return false;
    return true;
}

File: KalkJava/Kalk/Espressione.java
package Kalk;

public abstract class Espressione {
	protected String text;
	protected Result result;

	protected final boolean checkParentesi() {
		//Controlla il bilanciamento delle parentesi
		int check = 0;
		for(int i = 0; i < text.length(); ++i) {
			if(text.charAt(i) == '(') {
				++check;
			} else if(text.charAt(i) == ')' && --check < 0) {
				return false;
			}
		}
		if(check != 0) {
			return false;
		}
		for(int i = 0; i < text.length() - 1; ++i) {
			if(text.charAt(i) == '(' && text.charAt(i + 1) == ')') {
				return false;
			}
		}
		return true;
	}
	protected Espressione(String s) {
		this.text = s;
		this.result = null;
	}
	
	public abstract double toDouble();

	public final Result getResult() {
		return result;
	}

	public final String getText() {
		return (String)text;
	}

	public abstract Result calculate();
}